src/kms/Crypto.ts

```diff
@@ -252,3 +270,131 @@ export const decryptBatchDocuments = (
+        generateDeterministicEncryptedFieldHeader(primary.tenantSecretNumericId),
+        deterministicEncryptBytes(field.plaintextField, dekBytes)
+    ).map(([header, encryptedDoc]) => Buffer.concat([header, encryptedDoc]));
+    return encryptedFuture.map((encrypted) => ({
+        encryptedField: encrypted,
+        derivationPath: field.derivationPath,
+        secretPath: field.secretPath,
+    }));
+};
+
+/**
+ * Encode the tenant secret ID as 4 bytes, then attach two bytes of 0s. Fails if the tenant secret ID can't fit into 4 bytes.
+ */
+export const generateDeterministicEncryptedFieldHeader = (tenantSecretId: number): Future<TenantSecurityException, Buffer> => {
+    if (tenantSecretId < 0 || tenantSecretId > Math.pow(2, 32) - 1) {
+        return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_HEADER_ERROR, "Failed to generate header."));
+    }
+    const arr = new ArrayBuffer(4);
+    const view = new DataView(arr);
+    view.setUint32(0, tenantSecretId, false); // byteOffset = 0; litteEndian = false
+    return Future.of(Buffer.concat([Buffer.from(view.buffer), DETERMINISTIC_HEADER_PADDING]));
+};
+
+/**
+ * Encrypt the provided bytes with the provided key using AES-256-SIV.
+ */
+const deterministicEncryptBytes = (bytes: Buffer, key: Buffer): Future<TenantSecurityException, Buffer> => {
+    return Future.tryP(() => miscreant.SIV.importKey(key, "AES-SIV", cryptoProvider))
+        .flatMap((siv) =>
+            Future.tryP(async () => {
+                const encrypted = await siv.seal(bytes, []);
+                return Buffer.from(encrypted);
+            })
+        )
+        .errorMap((e) => new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_ENCRYPT_FAILED, e.message));
+};
+
+/**
+ * Decrypt the deterministically encrypted field using the associated tenant secrets contained in `derivedKeys`.
+ * If `reencrypt` is true, return a failure if the field was already encrypted to the current tenant secret.
+ */
+export const deterministicDecryptField = (
+    encryptedField: DeterministicEncryptedField,
+    derivedKeys: DerivedKey[] | undefined,
+    reencrypt = false
+): Future<TenantSecurityException, DeterministicPlaintextField> => {
+    const decryptedFuture = decomposeDeterministicField(encryptedField.encryptedField).flatMap((parts) => {
+        const key = derivedKeys?.find((id) => id.tenantSecretNumericId === parts.tenantSecretId);
+        const primaryKey = derivedKeys?.find((key) => key.primary);
+        if (key === undefined) {
+            return Future.reject(new TscException(TenantSecurityErrorCode.DETERMINISTIC_FIELD_DECRYPT_FAILED, "Failed deterministic decryption."));
+        } else if (reencrypt && key === primaryKey) {
+            return Future.reject(
+                new TscException(TenantSecurityErrorCode.DETERMINISTIC_REENCRYPT_SAME_SECRET, "Field is already encrypted to current tenant secret.")
+            );
+        } else {
```

Comment:
I think I agree; this was something from Colt's original design that he asked for (with a separate error code that they could handle as they want). But we should maybe discuss this in a call to make sure we're all good with it.

Until then, I'll move the check into its own function like you suggested; you're right that I was doing it here to avoid doing the decomposition and the `.find` twice, but you're also right that they should both be cheap, so I'll keep working on my premature optimization habit.